Add legend to calendar

Moliholy opened the pull request:
Closes #6105.
Closes #6106.
Closes #6108.
Closes #6128.
Closes #6129.
Closes #6107.

### Final result:


<img width="1049" alt="Screenshot 2024-01-17 at 11 52 39" src="https://github.com/indico/indico/assets/2722756/0a3aa607-0118-45ed-aa29-f886f9fde6ab">
<img width="973" alt="Screenshot 2024-01-17 at 11 52 31" src="https://github.com/indico/indico/assets/2722756/d6855ee0-a834-461d-86d5-945ab1dd9f8c">
<img width="878" alt="Screenshot 2024-01-17 at 11 52 25" src="https://github.com/indico/indico/assets/2722756/fc14a485-e89e-4a67-95e0-e896a7766cf6">
<img width="1641" alt="Screenshot 2024-01-17 at 11 52 09" src="https://github.com/indico/indico/assets/2722756/6d434673-3db9-44ac-9960-0f22707457d2">
<img width="1347" alt="Screenshot 2024-01-17 at 11 53 39" src="https://github.com/indico/indico/assets/2722756/ff8fa7de-8c3c-4324-b5ad-cd95a6e25e5e">

ThiefMaster commented on indico/modules/categories/controllers/display.py:
```diff
@@ -567,6 +569,10 @@ def _process(self):
 
 
 class RHCategoryCalendarViewEvents(RHDisplayCategoryBase):
+    class GroupBy(Enum):
+        category = 1
+        location = 2
```
```python
class GroupBy(IndicoStrEnum):
    category = auto()
    location = auto()
```

Moliholy replied:
Fixed in [6ffa0f9](https://github.com/indico/indico/pull/6110/commits/6ffa0f92dd8d56120b9cb3fcee140b2d1053ff15), and thanks again :)

Moliholy commented on indico/modules/categories/client/js/components/CalendarLegend.module.scss:
```diff
@@ -0,0 +1,36 @@
+// modify it under the terms of the MIT License; see the
+// LICENSE file for more details.
+
+body {
```
Totally true, forgot to remove that after testing it worked. Fixed in [6ffa0f9](https://github.com/indico/indico/pull/6110/commits/6ffa0f92dd8d56120b9cb3fcee140b2d1053ff15).

Moliholy commented on indico/modules/categories/controllers/display.py:
```diff
@@ -575,14 +581,23 @@ def _process_args(self):
+        try:
+            self.group_by = self.GroupBy[group_by]
+        except KeyError as e:
+            raise BadRequest(str(e))
```
Much clearer way, indeed. Fixed in [6ffa0f9](https://github.com/indico/indico/pull/6110/commits/6ffa0f92dd8d56120b9cb3fcee140b2d1053ff15), and thanks a lot for the tip.

Moliholy commented:
@ThiefMaster one thing remains, but I'm not sure this PR should cover it: the color palette does not look sufficient in case there are many locations or categories and in my opinion it should be expanded. Just drop this here to know what you think.

Moliholy commented:
Just updated the description by adding screenshots of the final result.

ThiefMaster commented:
for the palette: feel free to come up with a larger one or find some way to generate a nice one on the fly

ThiefMaster commented on indico/modules/categories/client/js/calendar.js:
```diff
@@ -74,19 +86,87 @@ import {$T} from 'indico/utils/i18n';
+            calendar.refetchEvents();
+          };
+          ReactDOM.render(
+            React.createElement(CalendarLegend, {items, groupBy: data.group_by, onFilterChanged}),
```
you can use jsx syntax in .js files as well, so no need to manually calls `React.createElement()`

Moliholy replied:
Thanks for the tip! Fixed in [c16abfa](https://github.com/indico/indico/pull/6110/commits/c16abfad4e60f93ed7c3db25c0f5a84fbdc402f0).

ThiefMaster commented on indico/modules/categories/controllers/display.py:
```diff
@@ -581,8 +590,12 @@ def _process(self):
-                 .options(load_only('id', 'title', 'start_dt', 'end_dt', 'category_id')))
+                 .options(load_only('id', 'title', 'start_dt', 'end_dt', 'category_id', 'own_venue_id')))
         events = self._get_event_data(query)
+        raw_categories = Category.query.options(load_only('id', 'title')).all()
```
you can't do that ;) it's a huge list (over 10k in the CERN instance) which will cause a huge slowdown when processing it in the JS code!

however, using the `detailed_category_chain` (make sure to undefer it to avoid O(n) queries) of each event should allow you to get rid of this altogether

Moliholy replied:
Oh, didn't know that. Fixed in [0841c74](https://github.com/indico/indico/pull/6110/commits/0841c74514dfba918707224da94aa199e184a984).

Moliholy commented:
> for the palette: feel free to come up with a larger one or find some way to generate a nice one on the fly

@ThiefMaster done in [87b248f](https://github.com/indico/indico/pull/6110/commits/87b248f4a0b29720c4baf47c650445b8af3f2d30). Here's the result:

<img width="1633" alt="Screenshot 2024-01-10 at 20 49 19" src="https://github.com/indico/indico/assets/2722756/678c5241-67fe-428f-9450-d5a5d08a9229">

ThiefMaster commented:
the greens of 3 and 4 look quite similar

Moliholy commented:
> the greens of 3 and 4 look quite similar

Yes, that can happen. The list is randomly generated with a given seed, so what I could try is to find a seed that creates no "collisions" on the first ~10 elements. Would that suffice?

ThiefMaster commented:
having something stable would certainly be good. full random is weird because reloading the page shouldn't give you new colors.

i'd be surprised if there isn't some nice logic to generate a random-ish palette without being fully random...

Moliholy commented:
> full random is weird because reloading the page shouldn't give you new colors.

It does't! It generates the exact same colors for a given category or location IDs, but it so happens that, since colors are infinite, there can be collisions for randomness. That's why I was suggesting to change the seed. Look here:
https://github.com/indico/indico/pull/6110/commits/87b248f4a0b29720c4baf47c650445b8af3f2d30#diff-791a87a96ed182d72dbd69f1d61092083942dd060d291db0b1611abe5e84e729R64

Say I use the following:

```python
random.seed(BASE_SEED + seed)
```

Where `BASE_SEED` is an `int` that does not create collisions with the first ~10 categories or locations. If I were to find that number, would it suffice?

ThiefMaster commented:
you'd need to [instantiate a separate RNG](https://docs.python.org/3/library/random.html#random.Random) to avoid messing with randomness in other places (which would no longer be random when setting a global seed).

anyway, what i was thinking was not going fully random but e.g. doing increments on the hue first (that gives you very different colors), then varying on the saturation/brightness to get shades.

That way you first get very very distinct colors, and then variations of those colors. Even with just 20 hues and 3 brightnesses, you get 60 colors without similarities. add 4 variations of the saturation and you're at 240 - so probably more than you'll ever need.

(i'm too lazy to write a script to test this, but should be pretty straightforward to generate some html to show a bunch of colored squares using that logic to see if it's nice or not)

Moliholy commented:
Just applied your suggestion in [3ff4c61](https://github.com/indico/indico/pull/6110/commits/3ff4c61ba27cbb8ac896ce9557fdd1988f18e7be). Here's the result:

<img width="1626" alt="Screenshot 2024-01-12 at 12 57 50" src="https://github.com/indico/indico/assets/2722756/5fdcfd16-5859-4e0d-8158-20d4d88a0cdf">

Does it look better?

ThiefMaster commented on indico/modules/categories/controllers/display.py:
```diff
@@ -620,21 +621,34 @@ def _process(self):
+                    return category_data
+                else:
+                    return category_chain[index + 1]
+        return {'id': 0, 'title': 'No category'}
```
`0` is the root (Home) category, so not a good default one. however, you will never get an unlisted event in this view, so i don't think "No category" needs to be handled here (so i'd rather fail with an exception in that case)

Moliholy replied:
Fixed in [8a0afb7](https://github.com/indico/indico/pull/6110/commits/8a0afb730401a053eff6c593e5fb82a8d72ef430).

Moliholy commented:
@ThiefMaster PR description updated with screenshots.

ThiefMaster commented:
thanks! BTW, git commit messages should start with an uppercase char. not really an issue though since i'll eventually squash+merge it anyway

ThiefMaster reviewed (commented):
The code comments are just for information. I think you can remove all this, and simply add this to `WPCategoryCalendar` instead:

```python
    def _get_breadcrumbs(self):
        return render_breadcrumbs(_('Calendar'), category=self.category)
```

Making the last item of a breadcrumb trail clickable seems a bit pointless since you are already on that page, but if you really want to do that (and have a good reason for it), you can pass `Breadcrumb(title, url)` instead of just the title to `render_breadcrumbs`.

Moliholy commented:
> The code comments are just for information. I think you can remove all this, and simply add this to `WPCategoryCalendar` instead:

Sorry @ThiefMaster, but I don't know which part of the code you mean concretely. As for the suggestion itself, I don't quite like the title appearing on the right. Still, as you say:


<img width="556" alt="Screenshot 2024-01-18 at 17 56 19" src="https://github.com/indico/indico/assets/2722756/4606ed14-be1e-4c39-a544-8e7a63345195">

Moliholy commented on indico/web/breadcrumbs.py:
```diff
@@ -64,7 +65,7 @@ def render_breadcrumbs(*titles, category=None, event=None, management=False, cat
             category_url_factory = lambda cat, management: (url_for('categories.manage_content', cat)
                                                             if management and cat.can_manage(session.user)
-                                                            else cat.url)
+                                                            else url_for(endpoint, cat) if endpoint else cat.url)
```
Yep, much cleaner that way. Done in [d584678](https://github.com/indico/indico/pull/6110/commits/d584678e5382de0a53f9c9165590d7bde0b72850).

ThiefMaster replied:
Oh I see what you want to do now... you want clicking a category breadcrumb to go to the calendar view instead of the display view!

Moliholy replied:
Yes, but a much simpler way just occurred to me: [80c0555](https://github.com/indico/indico/pull/6110/commits/80c05555e27318d4f9bef08d29dbaa81a488ab8f). I honestly don't know how I didn't see it before...

Moliholy commented on indico/web/breadcrumbs.py:
```diff
@@ -44,6 +44,7 @@ def render_breadcrumbs(*titles, category=None, event=None, management=False, cat
                                  category url will be used.
     :param management: Whether the event/category breadcrumb trail
                        should link to management pages.
+    :param section: section of the category to inspect. For example: calendar
```
Yes, that corresponds to my initial approach which in the end I discarded, forgetting to adapt this in the process. Fixed in [d584678](https://github.com/indico/indico/pull/6110/commits/d584678e5382de0a53f9c9165590d7bde0b72850).

ThiefMaster commented:
```
Warning: Failed prop type: withStyles(SingleDatePicker): unknown props found: yearsBefore, yearsAfter
withStyles(SingleDatePicker)
SingleDatePicker@https://indico3.eluvian.cern.ch/dist/js/common.479c6b5e.bundle.js:19425:5 [react.f6742c2d.bundle.js:85661:30](https://indico3.eluvian.cern.ch/dist/js/react.f6742c2d.bundle.js)
Warning: Failed prop type: SingleDatePicker: unknown props found: yearsBefore, yearsAfter
SingleDatePicker@https://indico3.eluvian.cern.ch/dist/js/react.f6742c2d.bundle.js:17457:18
WithStyles@https://indico3.eluvian.cern.ch/dist/js/react.f6742c2d.bundle.js:84006:27
withStyles(SingleDatePicker)
MediaQuery@https://indico3.eluvian.cern.ch/dist/js/react.f6742c2d.bundle.js:69617:20
_SizeSpec@https://indico3.eluvian.cern.ch/dist/js/common.479c6b5e.bundle.js:28236:11
SingleDatePicker@https://indico3.eluvian.cern.ch/dist/js/common.479c6b5e.bundle.js:19425:5
```

I think you need to add your new props to the `PROP_BLACKLIST` to avoid them being passed to the date picker component.

ThiefMaster commented on indico/modules/categories/client/js/components/CalendarLegend.jsx:
```diff
@@ -0,0 +1,90 @@
+
+function LegendItem({title, color, textColor, checked, url, isSpecial, onChange}) {
+  return (
+    <div styleName="legend-item">
```
This whole thing has a pointer cursor and feels clickable, but you have to really click the checkbox itself to toggle it...

Moliholy replied:
I removed the hover from the CSS.

ThiefMaster replied:
Having to click the checkbox itself is still poor UX, regardless of the hover effect... :/

Moliholy replied:
I managed to handle onClick events over the entire div, so now it should be more usable.

ThiefMaster commented:
The datepicker changes seem to affect other places where this component is being used, e.g. room booking:

![image](https://github.com/indico/indico/assets/179599/63a79c03-aebe-41ba-9de6-a9bf8f9fae4d)

Also, I think there's some CSS missing to format the date picker nicely?

![image](https://github.com/indico/indico/assets/179599/9426a4fa-047a-4e82-ad9a-4027abcc7757)


Or is the CSS to center the day numbers something custom we added?

ThiefMaster commented on indico/modules/categories/client/js/calendar.js:
```diff
@@ -100,14 +217,24 @@ import {$T} from 'indico/utils/i18n';
+    ReactDOM.render(
+      <SingleDatePicker
+        onDateChange={date => calendar.gotoDate(date.toDate())}
+        placeholder={Translate.string('Go to date...')}
```
It's *really* weird that this thing looks like an input field and not like a button

![image](https://github.com/indico/indico/assets/179599/eae7eb36-01e7-48ca-aa30-af65790d0e91)

Moliholy replied:
So, this is something kinda hard to explain... it seems the react-dates component ONLY can render itself with an input. On the other hand... fullcalendar.js ONLY allows to customize it by using buttons. So, what I did is mixing both: there's a rendered button that I managed to hide with CSS and at the same time I used the only component that the current datepicker is allowed to use.

I'm aware this is a dirty hack, but it's the best approach I managed to get. If we'd like to dive deeper we'd have to use another react component that allows more flexibility.

Moliholy replied:
I don't think so. Take a look at this: https://react-dates.github.io/react-dates/?path=/story/singledatepicker-sdp--default

There's no way to render the component without an input field.

Similar for fullcalendar's custom buttons: https://fullcalendar.io/docs/customButtons

ThiefMaster replied:
https://react-dates.github.io/react-dates/?path=/story/daypicker--single-month

ThiefMaster commented:
The alignment of the checkbox compared to the hover background looks strange:

![image](https://github.com/indico/indico/assets/179599/cf4570ed-8bb3-4613-9cf7-b98a94c97901)

ThiefMaster commented:
Shouldn't clicking a category in the legend open the calendar on the same date/view?

ThiefMaster commented on indico/modules/categories/client/js/components/CalendarLegend.jsx:
```diff
@@ -0,0 +1,90 @@
+
+import PropTypes from 'prop-types';
+import React, {useState} from 'react';
+import './CalendarLegend.module.scss';
```
move imports with side effects to the end of the import list (they cannot be auto-sorted)

Moliholy replied:
Fixed

ThiefMaster commented on indico/modules/categories/controllers/display.py:
```diff
@@ -52,13 +56,23 @@
+        brightness += 0.5
+    r, g, b = colorsys.hsv_to_rgb(hue, saturation, brightness)
+    background_color = _rgb_to_hex((r, g, b))
+    text_color = _rgb_to_hex((0, 0, 0))
```
![image](https://github.com/indico/indico/assets/179599/949e50b7-54ad-4943-9a7b-6074e11863e2)

I think for some colors the text need to become white.

Moliholy replied:
Hmmmmmm the brightness is guaranteed to be greater or equal to 0.5. In such case I think it's better to use a black color, unless we come up with a different threshold for such value. As an approach I've chosen 0.7. This is the result:

<img width="1070" alt="Screenshot 2024-01-19 at 01 45 37" src="https://github.com/indico/indico/assets/2722756/b1b2fdde-9ad4-450b-83fb-4549193c8d5b">

<img width="1069" alt="Screenshot 2024-01-19 at 01 47 35" src="https://github.com/indico/indico/assets/2722756/fb1f1bc0-30d0-49e1-9108-fd7364e5ea3d">

ThiefMaster commented on indico/modules/categories/controllers/display.py:
```diff
@@ -567,22 +581,33 @@ def _process(self):
 
     def _process(self):
-        query = (Event.query
+        query = (
+            Event.query
```
Please keep the formatting style we use everywhere else in Indico.

That would be one of these:

```python
        query = (
            Event.query
            .filter(Event.starts_between(self.start_dt, self.end_dt),
                    Event.is_visible_in(self.category.id),
                    ~Event.is_deleted)
            .options(undefer(Event.detailed_category_chain),
                     load_only('id', 'title', 'start_dt', 'end_dt', 'category_id', 'own_venue_id'))
        )
```

```python
        query = (Event.query
                 .filter(Event.starts_between(self.start_dt, self.end_dt),
                         Event.is_visible_in(self.category.id),
                         ~Event.is_deleted)
                 .options(undefer(Event.detailed_category_chain),
                          load_only('id', 'title', 'start_dt', 'end_dt', 'category_id', 'own_venue_id')))
```

I think the second version is more common (and in fact it's what the `ongoing_events` query right below uses).

ThiefMaster commented:
oh and please rebase (there are no conflicts anyway)

Moliholy commented:
> The datepicker changes seem to affect other places where this component is being used, e.g. room booking:

Totally right, my apologies. I added [073ba79](https://github.com/indico/indico/pull/6110/commits/073ba792c6346acd8a2e383610c843599ab5a60f) to fix it.

<img width="1044" alt="Screenshot 2024-01-19 at 01 02 38" src="https://github.com/indico/indico/assets/2722756/4f6ea12c-14dc-460a-ad87-d2c08d5e597a">

Does it look better now?

ThiefMaster commented:
no merge commits please - they can be a huge pain in the ass to deal with. better to just `git pull --rebase origin master`

Moliholy commented:
> no merge commits please - they can be a huge pain in the ass to deal with. better to just git pull --rebase origin master

Understood. Thanks for the heads up 👍

> What makes the first category in that list "special"? Also it looks like the colored "icon" is getting squeezed when the category name is too long. flex-shrink: 0 fixes it.

Thanks for the `flex-shrink` suggestion. It is now applied. As for the "special", categories are sorted so that the category in which calendar you're in has two particularities:

1) It's shown in the legend as "This category" and has no hyperlink.
2) It is always sorted so that it's the first one.

I honestly don't know how you did it, but your case does not meet these requirements. It's definitely a bug, but I don't know how to reproduce it. Could you please paste what the server delivered to you?

> Shouldn't clicking a category in the legend open the calendar on the same date/view?

I agree with you, but we are clicking in a hyperlink, which implies we'd have to pass start and end dates as GET parameters. I'm not against, but I thought that'd be out of scope for this PR. The same could actually apply to the breadcrumbs, for instance, and in that case it wouldn't be so easy.

> The alignment of the checkbox compared to the hover background looks strange:

It's now fixed.

> The datepicker changes seem to affect other places where this component is being used, e.g. room booking:

This is now fixed

> Or is the CSS to center the day numbers something custom we added?

This is something being added in that particular screen. I didn't really look into it.

ThiefMaster commented:
The red one still has somewhat poor contrast 
![image](https://github.com/indico/indico/assets/179599/bffba05e-daac-4be1-8601-9dc465c14b4a)

ThiefMaster commented:
> As for the "special", categories are sorted so that the category in which calendar you're in has two particularities:
> - It's shown in the legend as "This category" and has no hyperlink.
> - It is always sorted so that it's the first one.

This was not the case in my example. I was in the Home category (which has no events by itself!), so *all* categories in the legend are subcategories. Maybe you are only handling cases where some events are directly in the current category?

ThiefMaster commented:
> I agree with you, but we are clicking in a hyperlink, which implies we'd have to pass start and end dates as GET parameters. I'm not against, but I thought that'd be out of scope for this PR. The same could actually apply to the breadcrumbs, for instance, and in that case it wouldn't be so easy.

I think for the breadcrumbs it's fine for now (even though it might not be too hard with some just slightly ugly JS magic to add a query string there was well), but for the legend I'm almost certain some user will consider it a bug as well...

ThiefMaster commented on indico/modules/categories/client/js/components/CalendarLegend.jsx:
```diff
@@ -0,0 +1,86 @@
+  ));
+  const options = [
+    {text: Translate.string('Category'), value: 'category'},
+    {text: Translate.string('Location'), value: 'location'},
```
I'd call this "Venue" since that's the term we use everywhere else as well? (same in the 'No location' text)

Moliholy replied:
Fixed

Moliholy commented:
> The red one still has somewhat poor contrast

That red has brightness equal to 1, so I'll directly use white text in all cases.

> This was not the case in my example. I was in the Home category (which has no events by itself!)

Oh yes, the corner case is when the category itself has no events! Nice catch!

> I think for the breadcrumbs it's fine for now (even though it might not be too hard with some just slightly ugly JS magic to add a query string there was well), but for the legend I'm almost certain some user will consider it a bug as well...

So... just to be clear: we don't need to implement that functionality (or at least for now). Am I correct?

Moliholy commented:
All issues addressed!

<img width="1639" alt="Screenshot 2024-01-19 at 13 34 24" src="https://github.com/indico/indico/assets/2722756/328559bf-c7ca-448e-823d-4afba7781cb8">

ThiefMaster commented:
![image](https://github.com/indico/indico/assets/179599/1de1fd30-7886-461e-8fe1-999aa1d77b9e)

I think you won't get around some actual logic to determine whether to use light or dark text...

Maybe the logic from something like https://github.com/Munawwar/apca-color-contrast would do the job? The few colors I entered on https://contrast-suggester.com/ were fine

ThiefMaster commented:
Clicking the button again or anywhere outside should close the date picker

ThiefMaster commented:
When changing the displayed period with the arrow buttons, opening the 'go to' date picker still stays in the current month. Would be better if it started showing the month currently shown.

Moliholy commented:
@ThiefMaster all issues addressed.

ThiefMaster commented on indico/modules/categories/client/js/calendar.js:
```diff
@@ -9,18 +9,69 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
 
+import CalendarSingleDatePicker from 'indico/react/components/dates/CalendarSingleDatePicker';
```
```suggestion
import {CalendarSingleDatePicker} from 'indico/react/components';
```

we re-export all those components from *components*

Moliholy replied:
Done 👍

ThiefMaster commented on indico/modules/categories/client/js/calendar.js:
```diff
@@ -9,18 +9,69 @@
+          text: Translate.string('Go to...'),
+          icon: 'i-button icon-calendar',
+          click: async (_, element) => {
+            if (!modalShown) {
```
`if (modalShown) {return}`

then you don't need to have all the code below indented

ThiefMaster replied:
also close the modal when clicking the button again

Moliholy replied:
It does close and and reopen again. I don't really find a good way to fix it, since the `onClose` handler is called before the button's `onclick` method.

A dirty hack that occurred to me would be to enclose the `modalShown = false` line inside a function that executes after a given timeout. What do you think?

ThiefMaster replied:
find a nicer fix for it (see my commit and explanation on matrix)

ThiefMaster commented on indico/modules/categories/controllers/display.py:
```diff
@@ -591,22 +594,41 @@ def _process(self):
+                else:
+                    return category_chain[index + 1]
+        # this should never happen
+        raise NotFound
```
`NotFound` is an HTTP error, it will be silent to instance admins (since it's just a 404 for the client). Better use a plain `Exception` for it (since that'll be logged as an error)

Moliholy replied:
Yep, that sounds good.

ThiefMaster commented on indico/modules/categories/controllers/base.py:
```diff
@@ -15,6 +15,7 @@
 
 class RHCategoryBase(RH):
     _category_query_options = ()
+    _section = ''
```
I think this is no longer needed

Moliholy replied:
Indeed 👍

ThiefMaster commented on indico/modules/categories/views.py:
```diff
@@ -65,6 +65,13 @@ class WPCategoryCalendar(WPCategory):
+        return render_breadcrumbs(
+            _('Calendar'),
+            category=self.category,
+            category_url_factory=lambda cat, management: url_for('categories.calendar', cat)
```
what about using the default display page for the current category? (since you're already on the calendar page for that one anyway)

Moliholy replied:
The default view would point to `categories.display`, and not to `categories.calendar`. That'd go against what was described in #6129:

> The subcategories listed in the legend should be links to the calendar view of those subcategories.

ThiefMaster replied:
This is for the breadcrumbs. What I mean is that if you are e.g. on "Home -> Calendar" then "Home" links to the category display page. If you are in "Home -> Test -> Calendar" then "Test" links to the category display page and "Home" links to the calendar view.

Moliholy replied:
That'd alter the functionality of items that apparently look similar. A user would expect an item displayed in a collection of similar ones to behave similarly to other of the same nature, but what you propose would change the behaviour of a given item, which looks misleading to me.

Not a strong opinion though, so as you prefer.

OmeGak replied:
I would find this behavior rather confusing. Why would the breadcrumb links send me to different views? And what is the benefit with the last segment of the breadcrumb sending me to the category display view? Especially when there is a big button saying "Back to category page".

ThiefMaster replied:
I especially found it with in case of "Home -> Calendar". But I had the same worry that it might be a bit strange to have different logic. So with 2.5 people thinking it's weird, let's not do it

ThiefMaster commented on indico/web/client/js/react/util/html.js:
```diff
@@ -26,10 +26,18 @@ export function toClasses(...params) {
  * the actual component.
  */
-export function injectModal(renderFunc) {
+export function injectModal(renderFunc, position) {
```
```suggestion
export function injectModal(renderFunc, position = undefined) {
```

it's nicer to be clear that this argument is optional

Moliholy replied:
Done

ThiefMaster commented on indico/web/client/js/react/util/html.js:
```diff
@@ -26,10 +26,18 @@ export function toClasses(...params) {
+    container.style.position = 'absolute';
+    container.style.left = `${left}px`;
+    container.style.top = `${top}px`;
+    container.style.zIndex = '1001';
```
not the biggest fan of hardcoding this hear, but i don't have a good idea on how to do it more nicely...

Moliholy replied:
Yep, I didn't like it either, but didn't manage to come up with something cleaner 😞

ThiefMaster commented:
It still breaks room booking ;/

![image](https://github.com/indico/indico/assets/179599/5d42f2c8-f66d-4e31-9b3a-5fe7d13864c3)
![image](https://github.com/indico/indico/assets/179599/6ba50919-d350-49b8-bbc9-f920a78b7e63)

Moliholy commented:
@ThiefMaster ready for another round.

ThiefMaster commented:
![image](https://github.com/indico/indico/assets/179599/62cfd885-a7dc-42da-9951-b7d818e26586)

Maybe default to 0..+1 or keep this even disabled by default? In Room Booking selecting past years makes no sense.

ThiefMaster commented on indico/modules/categories/client/js/calendar.js:
```diff
@@ -68,7 +68,7 @@ import CalendarLegend from './components/CalendarLegend';
               }
             };
-            injectModal(
+            return injectModal(
```
I don't think that return is needed

Moliholy replied:
`async` functions always return a promise. If this one is not returned then a dummy one will be allocated and implicitly returned, so I thought it'd be very slightly more efficient.

ThiefMaster replied:
I'd say it does not make much sense to return something that's not used by the caller.. it implies to someone reading it the return value is used for *something*

Moliholy replied:
I don't really agree with that, since functions should work regardless of the caller and what it's doing with the return type. Still, I added [d493d9e](https://github.com/indico/indico/pull/6110/commits/d493d9e731c52440282e93698a89978cab4c0f2c) with the changes.
